### PR TITLE
Delete BaseGroup class

### PR DESCRIPTION
**Pull request by flyusfly**

### CODE REVIEW

File: ban/core/models.py [hunk 1]
```diff
@@ -69,30 +69,25 @@ def postcodes_resource(self):
...
-
-class PostCode(BaseGroup):
+class PostCode(NamedModel):
     resource_fields = ['code', 'name', 'municipality']
+
     code = db.PostCodeField(index=True)
+    municipality = db.ForeignKeyField(Municipality, related_name='postcodes')
 
     class Meta:
+        abstract = True
```
Comment (yohanboniface) [inline ban/core/models.py]: That's not an abstract class ;)
abstract means no table in the db.

File: ban/core/models.py [hunk 2]
```diff
@@ -169,7 +175,7 @@ def compute_cia(self):
         if not get_fantoir:
             return None
         return compute_cia(str(self.parent.municipality.insee),
-                           get_fantoir(), self.number, self.ordinal)
+                           get_fantoir, self.number, self.ordinal)
```
Comment (yohanboniface) [inline ban/core/models.py]: That's not OK, `get_fantoir` is a callable.
As discussed in chat, you can totally remove the need for checking parent.get_fantoir: now the parent will always have this method.

**Comment by yohanboniface** [comment]
Also missing the test for the added `order_by`, as suggested here: https://github.com/BaseAdresseNationale/ban/pull/132#issuecomment-227747547 (I think my example is missing an explicit `ordinal` value for all the created housenumbers).

File: ban/core/models.py (hunk 2 above)
Comment (flyusfly) [inline ban/core/models.py]: 1. As discussed in chat, we should do another PR wich should fix get_fantoir().
2. Test_housenumber.test_get_housenumber_with_cia() was KO.
The only answer was to fix this test and push the code.
The real answer would be the 2nd PR.

File: ban/core/models.py [hunk 3]
```diff
@@ -108,14 +103,25 @@ class Group(BaseGroup):
...
+    municipality = db.ForeignKeyField(Municipality, related_name='groups')
+
+    class Meta:
+        abstract = True
+
     @property
     def tmp_fantoir(self):
         return '#' + re.sub(r'[\W]', '', unidecode(self.name)).upper()
 
+    @property
```
Comment (yohanboniface) [inline ban/core/models.py]: Ah, that's why get_fantoir was returning a string…
Why making that a dynamic property?

**Comment by flyusfly** [comment]
Don't you want the following sort:
```
pk | number | ordinal 
----+--------+---------
 30 | 4      | 
 39 | 4      | BIS
(2 rows)
```?

**Comment by flyusfly** [comment]
Misunderstanding, sorry...

In postgres, nulls are listed at the end of a sort except if we have to say it clearly in the order by clause.
Example:

select pk, number, ordinal
from housenumber
where parent_id = 119
and number = '4'
order by number ASC NULLS FIRST, ordinal ASC NULLS FIRST;

 pk | number | ordinal 
----+--------+---------
 30 | 4      | 
 39 | 4      | BIS
(2 rows)

**Comment by yohanboniface** [comment]
> Don't you want the following sort:
pk | number | ordinal
----+--------+---------
30 | 4 |
39 | 4 | BIS
(2 rows)

Yes, and I want:

| pk | number | ordinal
| --- | ---------- | -------
| 30 | 4 |
| 39 | 4 | A
| 39 | 4 | B

This needs `ORDER BY ORDINAL ASC NULLS FIRST` :)

**Comment by yohanboniface** [comment]
In Peewee, I think we need to do it by hand, something like:

```
qs.select().order_by(SQL('ordinal NULLS FIRST'))
```

File: ban/tests/http/test_housenumber.py [hunk 4]
```diff
@@ -35,11 +35,26 @@ def test_get_housenumber_with_unknown_id_is_404(get, url):
...
+
+
+def test_add_district_to_housenumber():
+    housenumber = HouseNumberFactory(number="1", ordinal="A",
+                                     parent__municipality__insee="27638",
+                                     parent__fantoir="276380011")
+    postcode = PostCodeFactory(municipality=housenumber.parent.municipality)
+    hn2 = HouseNumberFactory(postcode=postcode, number="2", ordinal="")
+    hn1 = HouseNumberFactory(postcode=postcode, number="1", ordinal="")
+    hn2bis = HouseNumberFactory(postcode=postcode, number="2", ordinal="bis")
```
Comment (yohanboniface) [inline ban/tests/http/test_housenumber.py]: Also add one with `ŧer` to be sure ordinal are ordered the way we want.

File: ban/tests/http/test_housenumber.py [hunk 5]
```diff
@@ -35,11 +35,26 @@ def test_get_housenumber_with_unknown_id_is_404(get, url):
...
+                                     parent__municipality__insee="27638",
+                                     parent__fantoir="276380011")
     resp = get(url('housenumber-resource', id=housenumber.cia,
                    identifier="cia"))
     assert resp.status == falcon.HTTP_200
-    assert resp.json['number'] == "22"
+    assert resp.json['number'] == "1"
+
+
+def test_add_district_to_housenumber():
```
Comment (yohanboniface) [inline ban/tests/http/test_housenumber.py]: I don''t think this is the name you wanted ;)

File: ban/tests/http/test_housenumber.py [hunk 6]
```diff
@@ -35,11 +35,26 @@ def test_get_housenumber_with_unknown_id_is_404(get, url):
...
+    housenumber = HouseNumberFactory(number="1", ordinal="A",
```
Comment (yohanboniface) [inline ban/tests/http/test_housenumber.py]: Why do we need that housenumber? And why setting explicitly the fantoir/insee?

File: ban/tests/http/test_housenumber.py (hunk 5 above)
Comment (yohanboniface) [inline ban/tests/http/test_housenumber.py]: Also: this is a model test, it should be in the `http` section of the tests.

File: ban/tests/http/test_housenumber.py [hunk 7]
```diff
@@ -35,11 +35,26 @@ def test_get_housenumber_with_unknown_id_is_404(get, url):
...
+                                     parent__municipality__insee="27638",
+                                     parent__fantoir="276380011")
+    postcode = PostCodeFactory(municipality=housenumber.parent.municipality)
+    hn2 = HouseNumberFactory(postcode=postcode, number="2", ordinal="")
+    hn1 = HouseNumberFactory(postcode=postcode, number="1", ordinal="")
+    hn2bis = HouseNumberFactory(postcode=postcode, number="2", ordinal="bis")
+    assert postcode.housenumbers[0] == hn1
+    assert postcode.housenumbers[1] == hn2
+    assert postcode.housenumbers[2] == hn2bis
```
Comment (yohanboniface) [inline ban/tests/http/test_housenumber.py]: We also need some tests that does the same checks for the housenumber<=>Group relations (parent, ancestors, mixed).
